### PR TITLE
Improve Puppet Forge badges

### PR DESCRIPTION
**Pull request by raphink**
This PR:

* Rebases the Puppet Forge code on the v3 API
* Retrieves additional infos from the Forge API:

  - 'dt': total downloads of a module
  - 'e': endorsement of a module (approved/supported)
  - 'f': feedback score (percentage)
* Add user statistics from the Forge API:

  - 'rc': user's release count
  - 'mc': user's module count

### CODE REVIEW

**Comment by raphink** [comment]
@espadrine  any news on this?

File: server.js [hunk 1]
```diff
@@ -2257,26 +2259,121 @@ cache(function(data, match, sendBadge, request) {
...
+          badgeData.colorscheme = 'lightgrey';
+          sendBadge(format, badgeData);
+          return;
+        }
+        var versions = releases.map(function(version) {
+          return version.version;
+        });
+        var version = latestVersion(versions);
+        if (unstable(version)) {
+          badgeData.colorscheme = "yellow";
```
Comment (espadrine) [inline server.js]: The color for unstable versions is orange elsewhere; could you use this here?

File: server.js [hunk 2]
```diff
@@ -2257,26 +2259,121 @@ cache(function(data, match, sendBadge, request) {
...
+        } else {
+          badgeData.colorscheme = "blue";
+        }
+        badgeData.text[1] = "v" + version;
```
Comment (espadrine) [inline server.js]: We can use `versionColor()` here, and fetch the version name directly (the `v` is appended in there).

File: server.js [hunk 3]
```diff
@@ -2257,26 +2259,121 @@ cache(function(data, match, sendBadge, request) {
...
+        if (total === 0) {
+          badgeData.colorscheme = 'red';
+        } else if (total < 100) {
+          badgeData.colorscheme = 'yellow';
+        } else if (total < 1000) {
+          badgeData.colorscheme = 'yellowgreen';
+        } else if (total < 10000) {
+          badgeData.colorscheme = 'green';
+        } else {
+          badgeData.colorscheme = 'brightgreen';
```
Comment (espadrine) [inline server.js]: We can use `downloadCountColor()` here.

File: server.js [hunk 4]
```diff
@@ -2257,26 +2259,121 @@ cache(function(data, match, sendBadge, request) {
...
+        }
+        badgeData.text[1] = metric(total) + ' total';
+      } else if (info === 'e') {
+        var endorsement = json['endorsement'];
```
Comment (espadrine) [inline server.js]: Small nit: `json.endorsement` is more in style.

File: server.js [hunk 5]
```diff
@@ -2240,14 +2240,16 @@ mapNugetFeed('myget\\/(.*)', 1, function(match) {
...
+  var info = match[1]; // either `v`, `dt`, `e` or `f`
+  var user = match[2];
+  var module = match[3];
+  var format = match[4];
   var options = {
     json: true,
-    uri: 'https://forge.puppetlabs.com/api/v1/releases.json?module=' + userRepo
+    uri: 'https://forgeapi.puppetlabs.com/v3/modules/'+user+'-'+module
   };
   var badgeData = getBadgeData('puppetforge', data);
```
Comment (espadrine) [inline server.js]: The badge tag should match the contents of the information. Could you use `puppetforge` for the version, but `downloads` for the downloads, `endorsement`, and `feedback`?

File: server.js [hunk 6]
```diff
@@ -2257,26 +2259,121 @@ cache(function(data, match, sendBadge, request) {
...
+camp.route(/^\/puppetforge\/([^\/]+)\/([^\/]+)\.(svg|png|gif|jpg|json)$/,
+cache(function(data, match, sendBadge, request) {
+  var info = match[1]; // either `rc` or `mc`
+  var user = match[2];
+  var format = match[3];
+  var options = {
+    json: true,
+    uri: 'https://forgeapi.puppetlabs.com/v3/users/'+user
+  };
+  var badgeData = getBadgeData('puppetforge', data);
```
Comment (espadrine) [inline server.js]: As above, the badge tag should reflect what each information contain. `releases` and `modules`, perhaps? Having a number on the right lets people know what they mean.

File: server.js [hunk 7]
```diff
@@ -2257,26 +2259,121 @@ cache(function(data, match, sendBadge, request) {
...
+        } else if (releases < 10) {
+          badgeData.colorscheme = 'yellow';
+        } else if (releases < 50) {
+          badgeData.colorscheme = 'yellowgreen';
+        } else if (releases < 100) {
+          badgeData.colorscheme = 'green';
+        } else {
+          badgeData.colorscheme = 'brightgreen';
+        }
+        badgeData.text[1] = metric(releases) + ' releases';
```
Comment (espadrine) [inline server.js]: Given a specific tag, having `' releases'` becomes redundant.

File: server.js [hunk 8]
```diff
@@ -2257,26 +2259,121 @@ cache(function(data, match, sendBadge, request) {
...
+        } else if (modules < 5) {
+          badgeData.colorscheme = 'yellow';
+        } else if (modules < 10) {
+          badgeData.colorscheme = 'yellowgreen';
+        } else if (modules < 50) {
+          badgeData.colorscheme = 'green';
+        } else {
+          badgeData.colorscheme = 'brightgreen';
+        }
+        badgeData.text[1] = metric(modules) + ' modules';
```
Comment (espadrine) [inline server.js]: Given a specific tag, having `' releases'` becomes redundant.

**Comment by espadrine** [comment]
Thanks a lot for the work you did, and I'm awfully sorry for letting this linger. Intense month at work.

File: server.js (hunk 1 above)
Comment (raphink) [inline server.js]: I actually left this code untouched. I don't mind changing it, but that will modify the current behaviour of the puppetforge badge.

File: server.js (hunk 5 above)
Comment (raphink) [inline server.js]: I don't mind doing that. Do you prefer if I use different routes for each type of badge, or if I keep one route the way I do it now?

File: server.js (hunk 2 above)
Comment (raphink) [inline server.js]: Again, I didn't touch this code. I don't mind changing it at all of you would prefer.

File: server.js (hunk 3 above)
Comment (raphink) [inline server.js]: That's a good one. I might want to use it in other PRs too. Is there some kind of generic color function where one could pass the desired values for each color?

File: server.js (hunk 4 above)
Comment (raphink) [inline server.js]: Ack, I can change that.

**Comment by raphink** [comment]
It's alright. Thanks for reviewing :smile:

File: server.js (hunk 1 above)
Comment (espadrine) [inline server.js]: Mmh, I'm indecisive on this. Every other badge uses orange to indicate an unstable version, so conformity would be welcome. However, if there is something about the puppetforge community that gives meaning to that yellow color, I'll follow what the puppetforge community wants.

(Note: I failed to notice that difference while reviewing this: https://github.com/badges/shields/pull/171/files)

File: server.js (hunk 2 above)
Comment (espadrine) [inline server.js]: I'd love you to change it.

File: server.js (hunk 5 above)
Comment (espadrine) [inline server.js]: Whichever way you find most easy to read is welcome.

File: server.js (hunk 3 above)
Comment (espadrine) [inline server.js]: Oh, that would be great! There isn't at the moment, but you're welcome to do it. All those functions are defined at the bottom, around line 3040.

File: server.js (hunk 1 above)
Comment (raphink) [inline server.js]: I don't think there's is anything specific about puppet modules, so if it's on to change the behaviour, I'm in favour of a unified one.

File: server.js (hunk 3 above)
Comment (raphink) [inline server.js]: Gotcha

File: server.js (hunk 5 above)
Comment (raphink) [inline server.js]: I fixed that by amending `badgeData.text[0]` when necessary. That seemed much easier.

**Comment by raphink** [comment]
@espadrine  All fixed I think.

File: server.js (hunk 6 above)
Comment (raphink) [inline server.js]: Fixed using `badgeData.text[0]` when necessary.

File: server.js [hunk 9]
```diff
@@ -3134,25 +3193,21 @@ function fetchFromSvg(request, url, cb) {
...
-  }
+  floorCountColor(percentage, 80, 90, 100);
 }
 
 function downloadCountColor(downloads) {
-  if (downloads === 0) {
+  floorCountColor(downloads, 10, 100, 1000);
+}
+
+function floorCountColor(value, yellow, yellowgreen, green) {
```
Comment (raphink) [inline server.js]: Added a `floorCountColor()` generic function as promised.

**Comment by espadrine** [comment]
Thanks a lot!

**Pull request merged**